Fix kubectl annotate and label to use versioned objects when operating

soltysh opened the pull request:
Currently `kubectl annotate` and `label` uses the internal representation when posting changes to the server. In the light of proposed changes (#3933) and the fact that in OpenShift we already don't have json tags in internal representation we need to update both commands to use versioned objects.

@kubernetes/kubectl

soltysh committed:
Fix kubectl annotate and label to use versioned objects when operating

kargakis commented:
Same applies for `kubectl label`

kargakis commented:
cc: @liggitt

soltysh commented:
> Same applies for kubectl label

Didn't notice, I'll add that one as well.

deads2k commented on pkg/kubectl/cmd/annotate.go:
```diff
@@ -115,6 +119,13 @@ func NewCmdAnnotate(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 
 // Complete adapts from the command line args and factory to the data required.
 func (o *AnnotateOptions) Complete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string) (err error) {
+	o.outputVersion, err = cmdutil.OutputVersion(cmd, &registered.EnabledVersionsForGroup(api.GroupName)[0])
```
does this work right for things in the extensions group?

liggitt replied:
Could also panic if no versions are found for the group

kargakis replied:
right, let's avoid accessing indexes like that

deads2k commented on pkg/kubectl/cmd/annotate.go:
```diff
@@ -193,7 +204,11 @@ func (o AnnotateOptions) RunAnnotate() error {
 		}
 
-		name, namespace, obj := info.Name, info.Namespace, info.Object
+		obj, err := resource.AsVersionedObject([]*resource.Info{info}, false, o.outputVersion.String())
```
I would have expected you to use the version you negotiated with the API server so that when you submit the patch, your patch matches the version you're writing to.

liggitt replied:
Yeah, we had a long discussion about this in our env command.

soltysh replied:
I've used here the pattern we have in our env command specifically.

deads2k replied:
> I've used here the pattern we have in our env command specifically.



> Yeah, we had a long discussion about this in our env command.

This would seem to be a discussion I missed.  How can this be right?  I don't specify an output version.  Locally I have v1 and v2, v2 has priority.  v2 gets chosen, but v1 is the server-side level.  The patch is build using v2, which has labels under the tag `anything else`.  That patch gets sent to the v1 API and nothing happens.

liggitt replied:
I don't think this is right. Env built the patch using the server API version

soltysh replied:
Right, my bad, I was also looking at conver in kubectl. Will fix that later today.

soltysh commented:
@deads2k @liggitt addressed your comments.

thockin commented:
I am not the right assignee for this.  @janetkuo, can you claim this or find someone better suited to review it.

liggitt commented on pkg/kubectl/cmd/annotate.go:
```diff
@@ -193,7 +193,11 @@ func (o AnnotateOptions) RunAnnotate() error {
 		}
 
-		name, namespace, obj := info.Name, info.Namespace, info.Object
+		obj, err := resource.AsVersionedObject([]*resource.Info{info}, false, info.Mapping.GroupVersionKind.Version)
```
this is sort of going around our elbow to reach our nose... we're using the GroupVersionKind from the mapping, which is not what we've negotiated with the server. If we're just going to use the info the mapping, we should just use `info.VersionedObject.(runtime.Object)` instead

liggitt replied:
hmm, maybe not... looks like VersionedObject is nil when loading from the server

liggitt replied:
also, looks like AsVersionedObject attempts to use the version you give it, then falls back to `info.Mapping.GroupVersionKind.GroupVersion().String()`, so I guess (?) we can ask it to convert to the version we've negotiated with the server, and if it can't it'll convert to the versioned mapping?

liggitt replied:
I'd expect something more like https://github.com/soltysh/origin/commit/b93e90f57e9872a7e5b52a0b934a1c0d367d6cd4

smarterclayton commented:
Hrm, this overlaps with work in #17922 which covers some of the problem.
Can you double check that my changes don't overlap with yours?

On Fri, Jan 22, 2016 at 8:13 AM, k8s-merge-robot <notifications@github.com>
wrote:

> Labelling this PR as size/M
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19988#issuecomment-173917114>
> .
>

soltysh commented:
@smarterclayton will do on Monday

soltysh commented:
@smarterclayton it does not, it rather builds on top of your encoder you've added in kubectl commands.
@liggitt as mentioned I've rebased it on top of recent changes and updated according to your suggestions ptal.

deads2k commented:
rebase.  This has non-conflicting conflicts on the scheme/codec changes

soltysh commented:
@janetkuo ptal or assign someone as @thockin mentioned

bgrant0607 commented:
cc @kubernetes/kubectl @kubernetes/goog-csi

bgrant0607 commented:
@thockin In the future, feel free to assign to me and I'll delegate. I only just noticed this now.

soltysh commented:
@janetkuo ping

smarterclayton commented:
I just saw this, but I think the current way we're using AsVersionedObjects is broken and we need to move off it.  What I've done here: https://github.com/openshift/origin/pull/7331/files#diff-6dacd1007983d2583c28ce6fe4701ca5R135 is more correct.  If this code has tests we can merge it and come back in 1.3 and fix it, but I don't want to go much further in using AsVersionedObjects for this as it's not really correct.

soltysh commented:
@smarterclayton let's merge this as is and I'll work on fixing all the
places to the proposed approach as a follow -up.
On Feb 16, 2016 6:04 PM, "Clayton Coleman" <notifications@github.com> wrote:

> I just saw this, but I think the current way we're using
> AsVersionedObjects is broken and we need to move off it. What I've done
> here:
> https://github.com/openshift/origin/pull/7331/files#diff-6dacd1007983d2583c28ce6fe4701ca5R135
> is more correct. If this code has tests we can merge it and come back in
> 1.3 and fix it, but I don't want to go much further in using
> AsVersionedObjects for this as it's not really correct.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19988#issuecomment-184773332>
> .
>

lavalamp commented:
This would be a great use case for the versioned client @krousey has been working on.

lavalamp commented:
I don't know what "AsVersionedObject" is, but I'm willing to lgtm this as a temporary measure if some tests show up. Or you could do something that makes Clayton happy.

soltysh commented:
@lavalamp applied @smarterclayton suggestion ptal

smarterclayton commented:
This is fine for me

soltysh commented:
@k8s-bot unit test this github flake: #21484

lavalamp commented:
LGTM, thanks!

soltysh commented:
@k8s-bot unit test this github flake: #21484 

If the test will still fail I'll rebase my PR.
